File: day10/handle.rb
class Array
  def split_sorted
    dup, res = self, []
    while dup.size > 1 do
      index, tmp = 0, []
      while dup.include?(dup[index] + 1) do
        index += 1
      end
      (index + 1).times { tmp << dup.shift }
      res << tmp
    end

    res
  end
end

class Handle
  def initialize(file = 'data')
    @data = File.read(file).split("\n").map do |line|
      line.scan(/-?\d+/).map(&:to_i)
    end
  end

  # Part 1
  # Part 2
  def message_in_sky(data = @data)
    seconds = 0
    while true do
      seconds += 1
      points = data.map { |e| [e[0] + e[2] * seconds, e[1] + e[3] * seconds] }
      group_x = points.group_by(&:first).map { |k, v| [k, v.map(&:last)] }.to_h
      group_y = points.group_by(&:last).map { |k, v| [k, v.map(&:first)] }.to_h

      if is_msg?(group_x) || is_msg?(group_y)
        puts seconds
        print_msg points
        break
      end
    end
  end

  def is_msg?(group)
    res = group.keys.sort.split_sorted.map do |arr|
     arr.reduce([]) { |r, x| r + group[x] }
    end.map { |e| [e.min, e.max] }
    res.size > 1 && res.uniq.size == 1
  end

  def print_msg(points)
    xs, ys = points.map(&:first), points.map(&:last)
    (ys.min..ys.max).each do |y|
      (xs.min..xs.max).each do |x|
        if points.include? [x, y]
          print '#'
        else
          print ' '
        end
      end
      print "\n"
    end
  end
end

puts Handle.new.message_in_sky

File: 2018/day10/handle.exs
defmodule Message do
  def parse_inputs(file) do
    file
    |> File.stream!()
    |> Enum.map(fn line ->
      [[x], [y], [vx], [vy]] = Regex.scan(~r/(-?\d+)/, line, capture: :all_but_first)
      {{String.to_integer(x), String.to_integer(y)}, {String.to_integer(vx), String.to_integer(vy)}}
    end)
  end

  def step({{x, y}, {dx, dy}}), do: {{x + dx, y + dy}, {dx, dy}}

  def process(stars, seconds, true) do
    IO.inspect(seconds: seconds)

    stars
    |> Enum.map(fn {p, _} -> p end)
    |> print_msg()
  end
  def process(stars, seconds, false) do
    stars = Enum.map(stars, &step/1)
    process(stars, seconds + 1, is_msg(stars))
  end

  defp is_msg(stars) do
    points =
      stars
      |> Enum.map(fn {p, _} -> p end)
      |> Enum.sort(&(elem(&1, 0) < elem(&2, 0)))

    y_group_by_x = Enum.group_by(points, &elem(&1, 0), &elem(&1, 1))

    heights =
      points
      |> Enum.map(fn {x, _} -> x end)
      |> Enum.uniq()
      |> group_x()
      |> Enum.map(fn list ->
        list
        |> Enum.reduce([], fn x, acc ->
          acc ++ y_group_by_x[x]
        end)
      end)
      |> Enum.map(fn list ->
        Enum.max(list) - Enum.min(list)
      end)

    length(heights) > 1 && heights |> Enum.uniq() |> length() == 1
  end

  defp group_x(list) do
    list
    |> Enum.reduce([[]], fn x, acc ->
      acc
      |> Enum.at(0)
      |> Enum.at(0)
      |> case do
        nil ->
          [[x]]

        prx ->
          if x == prx + 1 do
            [p | n] = acc
            [[x | p] | n]
          else
            [[x] | acc]
          end
      end
    end)
  end

  defp print_msg(points) do
    dxs = Enum.map(points, fn {x, _} -> x end)
    dys = Enum.map(points, fn {_, y} -> y end)

    (Enum.min(dys)..Enum.max(dys))
    |> Enum.each(fn y ->
      (Enum.min(dxs)..Enum.max(dxs))
      |> Stream.map(fn x ->
        if Enum.any?(points, fn p -> p == {x, y} end) do
          "#"
        else
          "."
        end
      end)
      |> Enum.join("")
      |> IO.puts()
    end)

    IO.puts("\n")
  end
end

case System.argv() do
  [input_file] ->
    input_file
    |> Message.parse_inputs()
    |> Message.process(0, false)

  _ ->
    IO.puts(:stderr, "usage: elixir handle.exs filename")
end
